### DISCUSSION TITLE
Files S3 - JSON file type - zero bytes response

### DISCUSSION THREAD

**Discussion by emmems** [pocketbase/pocketbase#2068] [Q&A]
Hi! 👋 

I have pocketbase connected to S3 (local works flawesly) via R2 CloudFlare.

The situation is:
1. uploading file to database with extension json/gltf (application/json)
2. then I have link to it: http://localhost:8090/api/files/tdf4vk1tlytuzua/t3t2t8whz4fp41u/my_file_gHctbnx2m7.json
3. I'm doing this request:
```
GET http://localhost:8090/api/files/tdf4vk1tlytuzua/t3t2t8whz4fp41u/my_file_gHctbnx2m7.json
```
4. That is my response:
```
HTTP/1.1 200 OK
Accept-Ranges: bytes
Access-Control-Allow-Origin: *
Cache-Control: max-age=2592000, stale-while-revalidate=86400
Content-Disposition: attachment; filename=my_file_gHctbnx2m7.json
Content-Length: 0
Content-Security-Policy: default-src 'none'; media-src 'self'; style-src 'unsafe-inline'; sandbox
Content-Type: application/json
Last-Modified: Wed, 15 Mar 2023 08:49:29 GMT
Vary: Origin
X-Content-Type-Options: nosniff
X-Xss-Protection: 1; mode=block
Date: Wed, 15 Mar 2023 08:55:23 GMT

<Response body is empty>
```

Additionally:
👍 Images are working just fine 
it doesn't matter if it is localhost or deployed on server.

GUI logs are returning 200.

My setup:

- Go 1.20
```
github.com/pocketbase/dbx v1.10.0
github.com/pocketbase/pocketbase v0.13.4
```

If someone knows why it might not work (who is/was working on it) then let me know and I can fix it by myself 😄

**Comment by ganigeorgiev** [comment]
I'm not able to reproduce the issue and json files are uploaded and downloaded for me.

How did you upload the file?

Are you sure that your S3 settings are correct?

Does the file exist in the S3 storage?

Could you provide a minimal reproducible repo?

**Reply by emmems** [reply]
Wow I'm impressed how fast you replied man, wow really appreciate it! Thank you :) 

- How did you upload the file?
Just clicking "Upload new file", choosing file, "Save changes".

- Are you sure that your S3 settings are correct?
Green dot is there, images are working just fine so I assume it is okay

- Does the file exist in the S3 storage?
Yes, one of screenshots below shows that :) 

- Could you provide a minimal reproducible repo?
Yes, sure :) Here you are:

1. 
```
func main(){
	app := pocketbase.NewWithConfig(&pocketbase.Config{
		DefaultDebug:         false,
		DefaultDataDir:       "pb_data",
		DefaultEncryptionEnv: "./pb_encrypted/",
		HideStartBanner:      false,
		DataMaxOpenConns:     0,
		DataMaxIdleConns:     0,
		LogsMaxOpenConns:     0,
		LogsMaxIdleConns:     0,
	})

	if err := app.Start(); err != nil {
		log.Fatalln("Failed to start pocketbase:", err)
	}
}
```
2. run **go run main.go serve**
![CleanShot 2023-03-15 at 10 31 15](https://user-images.githubusercontent.com/30047262/225267483-e4651d7a-af8b-48e9-a9f4-28f7040be315.png)

3. create account

4. go to **file storage** and fill R2 Cloudflare data.
![CleanShot 2023-03-15 at 10 27 42](https://user-images.githubusercontent.com/30047262/225268045-0e788a04-53e1-4910-a3bb-01af4c7284b1.png)


5. create simple table -> upload .json file.
![CleanShot 2023-03-15 at 10 30 11](https://user-images.githubusercontent.com/30047262/225267574-f5e860de-45cc-40d0-a452-29e9a7da8b7e.png)
![CleanShot 2023-03-15 at 10 30 18](https://user-images.githubusercontent.com/30047262/225267638-6683cdf6-9dc9-4606-888b-83b70065d4ce.png)

That's it.

S3 screenshot:
![CleanShot 2023-03-15 at 10 30 45](https://user-images.githubusercontent.com/30047262/225268516-8fe03f08-5ca5-4751-b263-918d844cfed4.png)
- File downloaded directly from S3 is fine.

This is file returned by pocketbase:
![CleanShot 2023-03-15 at 10 39 07](https://user-images.githubusercontent.com/30047262/225269468-0a1a495f-aca0-4941-9d5c-05822764d477.png)


Weird situation I guess.

**Reply by ganigeorgiev** [reply]
Hm, it could be something specific to Cloudflare R2.

I've tested it once again with the Vultr and Wasabi S3 offerings and it is working fine for me.

Your example config looks OK, I'm just not sure what `./pb_encrypted/` is in the above (_also note that disabling the db pool limits could lead to excessive memory usage and "Too many open files" errors since with every db request it will open a new db connection_).

Is "Force path-style addrressing" necessary for Cloudflare R2?

Is the empty file content reproducible only with JSON files?

Is the error reproducible with the prebuilt executable?

**Comment by emmems** [comment]
Hi!

I'm sorry but I didn't have time to check that further but now I'm back!

I will firstly answer to questions:
- _Force path-style addrressing_ <- works same with/without this setting.
- _Is the empty file content reproducible only with JSON files?_ JSON/SVG as I discovered 
- _Is the error reproducible with the prebuilt executable?_ - yes, 0.14.0 as well.

To be honest I think this is not related to pocketbase itself but to library `"gocloud.dev/blob/s3blob"`.

Why? When I checked amazon s3 it works, linode works fine as well. I have checked nodejs s3 provider on cloudflare and it works fine as well.

The issue is here:
[/tools/filesystem/filesystem.go](https://github.com/pocketbase/pocketbase/blob/master/tools/filesystem/filesystem.go)
```
func (s *System) Serve(res http.ResponseWriter, req *http.Request, fileKey string, name string) error {
	br, readErr := s.bucket.NewReader(s.ctx, fileKey, nil)
```

where `br.Size()` returns somehow 0 size :/  if this is Cloudflare R2 and json/svg files.

**Reply by ganigeorgiev** [reply]
I'm not sure how to reproduce it.

> When I checked amazon s3 it works, linode works fine as well. I have checked nodejs s3 provider on cloudflare and it works fine as well.

I'm confused. Do you mean that AWS S3 and Linode works with PocketBase, but only R2 Cloudflare has the zero bytes issue? If that's the case, then maybe there is something with the Cloudflare S3 compatibility API that is causing it to fail.

**Reply by emmems** [reply]
Ohh it means:

linode S3, AWS S3- nodejs aws-s3 lib, pocketbase works ✅ 

cloudflare r2 - nodejs aws-s3 lib works ✅ 

cloudflare r2 - pocketbase doesn't work ✅ :/ it returns zero bytes.

So I'm assuming R2 has some issue with their compatibility API :/

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here - I've created a R2 account and was able to reproduce the above issue with the `aws-sdk-go` pacakge. Upgrading to `aws-sdk-go-v2` seems to resolve it (currently the changes are only in the [`develop` branch](https://github.com/pocketbase/pocketbase/commit/b9029010d9b33d19a2ebdbd0a5bc9199e1c71a14))._

**Reply by zknox72** [reply]
That is great timing. I was looking into using R2 for my project. And when can we expect these changes to be in next release
Thank you for your help 🙏

**Reply by jackbach** [reply]
i am having the same issue, running the latest pocketbase version (0.29.2)

zero bytes response on svg and json files

was this ever merged?
any know issues related to this?
any suggested fixes?

thank you!

**Reply by ganigeorgiev** [reply]
@jackbach The `aws-sdk-go-v2` was removed and replaced with a custom implementation for broader S3 vendors support - see https://github.com/pocketbase/pocketbase/discussions/6562. 

The new implementation is supposed to essentially replicate the same behavior as `aws-sdk-go-v2` minus the CRC checksum since they are not supported everywhere (even Cloudflare R2 had a note in their docs that they no longer were compatible with the latest version of the official AWS SDK and recommended downgrading at the beginning of this year; here  https://community.cloudflare.com/t/aws-s3-sdk-compatibility-inconsistencies-with-r2/759067 seems to say that it is "resolved" but I haven't really verified it).

It is possible that something else has broke their compatibility or that we are not sending back something correctly but I no longer have access to a test Cloudflare account and I'm not able to assist here.

If you think that it is a PocketBase issue, feel free to open a PR _(the S3 client code is located in https://github.com/pocketbase/pocketbase/tree/master/tools/filesystem/internal/s3blob/s3)_ and I'll try to have a look at it, but considering that it works fine with practically any other S3 offering such as Wasabi, Vultr, Flow, Minio, Backblaze, iDrive, etc., I think there is something specific with the Cloudflare R2 S3 compatibility implementation and I would suggest to report it to Cloudflare if they can have a look at it.

**Reply by jackbach** [reply]
thanks for the detailed answer!

as a workaround, i am serving assets directly from a public r2 bucket instead of proxying them through pocketbase.
that makes all assets public, which is fine by me.
i guess this also breaks thumbnails created "on the fly"? i am ok with that too.
is there something else i should be aware of?

**Reply by ganigeorgiev** [reply]
@jackbach I can't think of anything else but again I'd recommend to report the issue to Cloudflare as it is very likely to be something similar to the one that was in Flow (https://github.com/pocketbase/pocketbase/pull/6914#issuecomment-2956412046), e.g. incorrectly reported or missing `Content-Length` header in the S3 `getObject` API response.

**Comment by RayyanNafees** [comment]
@emmems How did u established the R2 connection ? I'm trying to do the same plz tell me if I missed a step

## Cloudflare Dashboard:
1. R2 Object Storage 
2. Create Bucket 
3. Go back to Manage API tokens 
4. Allow Admin Read & Write to all Buckets
5. Copy the Access key ID, Secret Access key, S3 endpoint

## Pocketbase File Setting
1. Set Endpoint to `https://24833a01ce01ca7114c835c251d54566.r2.cloudflarestorage.com`

2. Copy pasted the access key ID & Secret Access key

3. Region set to `auto`

4. bucket set to `bucket-name`

**Comment by yasmany1111** [comment]
For anyone running into this — the issue is that some S3 providers (Cloudflare R2 in particular) don't return a `Content-Length` header, or return `0`, in their `GetObject` response for certain content types like JSON, SVG, and text files. When that happens, `blob.Reader.Size()` returns `0`, so `http.ServeContent` thinks the file is empty and sends back nothing. Images work fine because their responses do include the correct `Content-Length`.

I ran into this on my own setup and patched it in my fork. The fix is straightforward: when the size comes back as `0` from `GetObject`, do a follow-up `HeadObject` request to get the real file size. This only fires when the header is actually missing, so there's no extra overhead in the normal case.

**New file: `tools/filesystem/internal/s3blob/size_resolver.go`**

```go
package s3blob

import (
	"context"

	"github.com/pocketbase/pocketbase/tools/filesystem/internal/s3blob/s3"
)

// resolveSize attempts to get the actual size of an object when
// Content-Length is missing or zero (common with some S3 providers
// that use chunked transfer encoding for certain content types).
func resolveSize(ctx context.Context, s3Client *s3.S3, key string, currentSize int64) int64 {
	if currentSize > 0 {
		return currentSize
	}

	headResp, err := s3Client.HeadObject(ctx, key)
	if err != nil {
		return currentSize
	}

	return headResp.ContentLength
}
```

**Modified: `tools/filesystem/internal/s3blob/s3blob.go` (`NewRangeReader`)**

```go
// Get the size, resolving via HEAD request if necessary
size := getSize(resp.ContentLength, resp.ContentRange)

// If size is 0 and not a zero-length read, try HEAD request
// (some S3 providers don't return Content-Length for certain content types)
if size == 0 && length != 0 {
    size = resolveSize(ctx, drv.s3, key, size)
}

return &reader{
    body: body,
    attrs: &blob.ReaderAttributes{
        ContentType: resp.ContentType,
        ModTime:     resp.LastModified,
        Size:        size,
    },
}, nil
```

If the HEAD request also fails for some reason, it just falls back to the original value — nothing breaks.

**Reply by ganigeorgiev** [reply]
At this point I'm almost confident that the issue is caused by a gzip compressed response. More details you can find in https://github.com/pocketbase/pocketbase/pull/7373#issuecomment-3617498036.

Or in other words - if you have control over the S3 server response, try to disable its gzip compression.

**Reply by yasmany1111** [reply]
The gzip point makes a lot of sense, I hadn't considered that the compressed response was what's throwing off the Content-Length.

I'll try disabling compression on my R2 setup and see if that resolves it on my end. Appreciate the continued work on the project.

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here - [v0.36.3](https://github.com/pocketbase/pocketbase/releases/tag/v0.36.3) added `Accept-Encoding: identity` for all S3 API requests which is supposed to instruct the vendor proxy/gateway to return noncompressed responses. There is no guarantee that it will be acknowledged or even forwarded in case multiple proxies are involved (GCS for example is known to alter the header) but another user in https://github.com/pocketbase/pocketbase/issues/7523 confirmed that it works correctly when used with R2._
